### ISSUE TITLE
multiprocessing.Event.wait hangs when interrupted by signal that sets the event

### ISSUE THREAD

**Issue by 71b9f251-89b5-4499-8262-f20064efd13c** [python/cpython#85772]
BPO | [41606](https://bugs.python.org/issue41606)
--- | :---
Files | <li>[event-test.py](https://bugs.python.org/file49419/event-test.py "Uploaded as text/plain at 2020-08-21.08:01:43 by salvatore.ingala"): Test code for multiprocessing.Event</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2020-08-21.08:01:43.342>
labels = ['type-bug', '3.8']
title = 'multiprocessing.Event.wait hangs when interrupted by signal that sets the event'
updated_at = <Date 2020-08-21.08:01:43.342>
user = 'https://bugs.python.org/salvatoreingala'
```

bugs.python.org fields:
```python
activity = <Date 2020-08-21.08:01:43.342>
actor = 'salvatore.ingala'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = []
creation = <Date 2020-08-21.08:01:43.342>
creator = 'salvatore.ingala'
dependencies = []
files = ['49419']
hgrepos = []
issue_num = 41606
keywords = []
message_count = 1.0
messages = ['375745']
nosy_count = 1.0
nosy_names = ['salvatore.ingala']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue41606'
versions = ['Python 3.8']
```

</p></details>

**Comment by marcos4758** [comment]
I just spent a lot of time with this problem. I have a multiprocessing.Event that sleeps the MainThread of a child process and an interrupt handler that sets that event. As described here, when the interrupt triggers the event, the process hangs. Doing some research on the web, including on the cited stackoverflow topic, I got the following solution:

Solutions for Linux with fork start method:
1.  In the interrupt handler, call ``threading.Thread(target=exit_event.set).start()``
2. Change multiprocessing.Event to threading.Event

Solution for Linux with spawn start method
1.  In the interrupt handler, call ``threading.Thread(target=exit_event.set).start()``

Windows:
1. Instead of using signal.signal() to change the interrupt handler, the solution was to use:
```
import win32api
win32api.SetConsoleCtrlHandler(handler, add)
```
Reference: https://docs.microsoft.com/en-us/windows/console/setconsolectrlhandler
This works like a charm. The key difference is that it runs the interrupt funcition on a thread (named Dummy-1).

2.  (Not a great solution) In the interrupt handler, call ``threading.Thread(target=exit_event.set).start()``
    - This solves the hang issue, but raises an InterruptedError exception
